File: 3211_generate_binary_strings_without_adjacent_zeros/Solution.scala
// LeetCode 3211 - Generate Binary Strings Without Adjacent Zeros
// https://leetcode.com/problems/generate-binary-strings-without-adjacent-zeros/

object Solution {
  def validStrings(n: Int): List[String] = {
    val ans = scala.collection.mutable.ListBuffer.empty[String]
    val t = new StringBuilder
    def dfs(i: Int): Unit = {
      if (i >= n) { ans += t.toString; return }
      var j = 0
      while (j < 2) {
        if ((j == 0 && (i == 0 || t.charAt(i - 1) == '1')) || j == 1) {
          t.append(('0' + j).toChar)
          dfs(i + 1)
          t.deleteCharAt(t.length - 1)
        }
        j += 1
      }
    }
    dfs(0)
    ans.toList
  }
}

File: 3211_generate_binary_strings_without_adjacent_zeros/solution.php
<?php
// LeetCode 3211 - Generate Binary Strings Without Adjacent Zeros
// https://leetcode.com/problems/generate-binary-strings-without-adjacent-zeros/

class Solution {
    private $n;
    private $ans;
    private $t;

    function validStrings($n) {
        $this->n = $n;
        $this->ans = [];
        $this->t = [];
        $this->dfs(0);
        return $this->ans;
    }

    private function dfs($i) {
        if ($i >= $this->n) { $this->ans[] = implode('', $this->t); return; }
        for ($j = 0; $j < 2; $j++) {
            if (($j === 0 && ($i === 0 || $this->t[$i - 1] === '1')) || $j === 1) {
                $this->t[] = (string)$j;
                $this->dfs($i + 1);
                array_pop($this->t);
            }
        }
    }
}
